typeid/typeid-go/typed/typed.go

```diff
@@ -0,0 +1,115 @@
+package typed
+
+import (
+	"fmt"
+
+	untyped "go.jetpack.io/typeid"
+)
+
+// IDType is an interface used to represent a statically checked ID type.
+// Example:
+// type UserIDType struct{}
+// func (UserIDType) Type() string { return "user" }
+type IDType interface {
+	Type() string
+}
+
+// TypeID is a unique identifier with a given type as defined by the TypeID spec
+type TypeID[T IDType] untyped.TypeID
+
+// New returns a new TypeID with a random suffix and the given type.
+func New[T IDType]() (TypeID[T], error) {
+	var prefix T
+	tid, err := untyped.New(prefix.Type())
+	if err != nil {
+		// Clients should ignore the id value when an error is present, but just
+		// in case, construct a "nil" id of the given type.
+		return Nil[T](), err
+	}
+	return (TypeID[T])(tid), nil
+}
+
+// Nil returns the null typeid of the given type.
+func Nil[T IDType]() TypeID[T] {
+	var prefix T
+	return TypeID[T](untyped.Must(untyped.From(prefix.Type(), "00000000000000000000000000")))
```

Comment:
Can this be 

```golang
return TypeID[T]{}
``` 

or maybe:
```golang
return TypeID[T]{suffix: untyped.NIL.Suffix()}
``` 

?

It would be missing the prefix but is that something that makes sense in an error call?  (If we are telling users to ignore the value, having it contain prefix data could be worse). In a similar vein, I assume initializing it with `00000000000000000000000000` (both here and untyped) is just so that functions that require a suffix value won't break, but is that a property we want when functions return non-nil errors.

Don't feel strongly and not a blocker, just sharing thoughts.